pyrolite/mineral/normative.py

```diff
@@ -153,213 +158,250 @@ def CIPW_norm(data, form="weight"):
+    mag_fltr = res["Fe2O3"] >= res["FeO"]
+    norm["magnetite"] = np.where(mag_fltr, res["FeO"], res["Fe2O3"])
+    res["Fe2O3"] = res["Fe2O3"] - norm["magnetite"]
+    res["FeO"] = res["FeO"] - norm["magnetite"]
+    norm["haematite"] = res["Fe2O3"] # assign any residual Fe2O3 to haeamatite
 
     # Subdivided normative minerals
-    data['MgFeO'] = data['MgO']+data['FeO']
-    data['Mg/Fe'] = data['MgO']/(data['FeO']+data['MgO'])
-    data['Fe/Mg'] = data['FeO']/(data['FeO']+data['MgO'])
-
-    # diopside,wollastonite, hypersthene
-    data['diopside'] = np.where(data['CaO']>=data['MgFeO'], data['MgFeO'], data['CaO'])
-    CaO = np.where(data['CaO']>=data['MgFeO'], data['CaO']-data['diopside'], 0)
-    MgFeO = np.where(data['CaO']>=data['MgFeO'], 0, data['MgFeO']-data['diopside'])
-    data['wollastonite'] = np.where(data['CaO']>=data['MgFeO'], CaO, 0)
-    data['hypersthene'] = np.where(data['CaO']>=data['MgFeO'], 0, MgFeO)
-    Y = np.where(data['CaO']>=data['MgFeO'], Y+(2*data['diopside'])+data['wollastonite'], Y+(2*data['diopside'])+data['hypersthene'])
-    data['CaO'] = CaO
-    data['MgFeO'] = MgFeO
-
+    res['MgFeO'] = res["MgO"] + res["FeO"]
+#     data['Mg/Fe'] = data['MgO']/(data['FeO']+data['MgO'])
+#     data['Fe/Mg'] = data['FeO']/(data['FeO']+data['MgO'])
+
+    # Diopside, Wollastonite, Hypersthene (provisional)
+    dio_flt = res["CaO"] >= res["MgFeO"] # where Mg and Fe are limiting
+    norm["diopside"] = np.where(dio_flt, res["MgFeO"], res["CaO"])
+    res["CaO"] = res['CaO'] - norm["diopside"]
+    res["MgFeO"] = res["MgFeO"] - norm["diopside"]
+    norm["wollastonite"] = res["CaO"] # assign residual calcium to wollastonite
+    norm["hypersthene"] = res["MgFeO"] # assign residual magnesium and iron to hypersthene
+    res["Y"] = res["Y"] + (2 * norm["diopside"]) + norm["wollastonite"] + norm["hypersthene"] # update silica sum
 
     # Quartz/undersaturated minerals
-    data['quartz'] = np.where(data['SiO2']>=Y, data['SiO2']-Y, 0)
-    D = np.where(data['SiO2']<Y, Y-data['SiO2'], 0)
-
-    # Olivine
-    olivine = np.where((D != 0) & (D<data['hypersthene']/2), D, 0)
-    olivine = np.where((D != 0) & (D>=data['hypersthene']/2), data['hypersthene']/2, olivine)
-    D1 = np.where((D != 0) & (D>=data['hypersthene']/2),D-(data['hypersthene']/2),0)
-    hypersthene = np.where((D != 0) & (D<data['hypersthene']/2), data['hypersthene']-(2*D), data['hypersthene'])
-    data['hypersthene'] = np.where((D != 0) & (D>=data['hypersthene']/2), 0, hypersthene)
-    data['olivine'] = olivine
+    qua_flt = res["SiO2"] >= res["Y"] # if silica is in excess
+    norm["quartz"] = np.where(qua_flt, res['SiO2']-res["Y"], 0)
+    res["D"] = np.where(qua_flt, 0, res["Y"] - res["SiO2"]) # amount of silica deficiency
+
+    # Olivine 
+    # Using two filters as we want to skip if deficiency = 0
+    oli_flt_1 = (res["D"] > 0) & (res["D"] < norm["hypersthene"]/2) 
+    oli_flt_2 = (res["D"] > 0) & (res["D"] >= norm["hypersthene"]/2)
```

Comment:
Bit clunky - maybe there is a more concise option

Reply:
You could do something with `np.max` here, but I think you'll still need some filters to update the silica deficiency etc.
```python
norm["olivine"] = np.vstack([res["D"], norm["hypersthene"]/2]).max(axis=0)
```
A bit clunky is OK for now, there's time for refining after it works as expected!

Also - do you need sequentially labelled intermediate deficiencies, or could you treat this the same as other components and modify a single column?